quadtree/quadtree.go

```diff
@@ -382,6 +420,93 @@ func (v *findVisitor) Visit(p geo.Pointer) {
 	return
 }
 
+type pointsQueueItem struct {
+	point    geo.Pointer
+	distance float64 // distance to point and priority inside the queue
+	index    int     // point index in queue
+}
+
+type pointsQueue []*pointsQueueItem
+
+func NewPointsQueue(capacity int) pointsQueue {
+	// We make capacity+1 because we need additional place for the greatest element
+	return make([]*pointsQueueItem, 0, capacity+1)
+}
+
+func (pq pointsQueue) Len() int { return len(pq) }
+
+func (pq pointsQueue) Less(i, j int) bool {
+	// We want pop longest distances so Less was inverted
+	return pq[i].distance > pq[j].distance
+}
+
+func (pq pointsQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *pointsQueue) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*pointsQueueItem)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *pointsQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1
+	*pq = old[0 : n-1]
+	return item
+}
+
+type kNearestVisitor struct {
+	point              *geo.Point
+	filter             Filter
+	k                  int
+	closest            pointsQueue
+	closestBound       *geo.Bound
+	maxAllowedDistance float64
+}
+
+func (v *kNearestVisitor) Bound() *geo.Bound {
+	return v.closestBound
+}
+
+func (v *kNearestVisitor) Point() *geo.Point {
+	return v.point
+}
+
+func (v *kNearestVisitor) Visit(p geo.Pointer) {
+	// skip this pointer if we have a filter and it doesn't match
+	if v.filter != nil && !v.filter(p) {
+		return
+	}
+
+	point := p.Point()
+	if d := point.SquaredDistanceFrom(v.point); d < v.maxAllowedDistance {
+		heap.Push(&v.closest, &pointsQueueItem{point: p, distance: d})
+		if v.closest.Len() > v.k {
+			heap.Pop(&v.closest)
+
+			// Actually this is a hack. We know how heap works and obtain top element without function call
+			top := v.closest[0]
+
+			v.maxAllowedDistance = top.distance
+
+			// We have filled queue, so we start to restrict searching range
+			d = math.Sqrt(top.distance)
+			x := v.point.X()
+			y := v.point.Y()
+			v.closestBound.Set(x-d, x+d, y-d, y+d)
+		}
+	}
+
+	return
```

Comment:
Is this `return` necessary?

Reply:
similar with previous `Visitor` implementation. Fixed both.